log-manager/src/main/java/io/airlift/log/LogHistoryManager.java

```diff
@@ -0,0 +1,207 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.airlift.log;
+
+import io.airlift.units.DataSize;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.PriorityQueue;
+import java.util.Set;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+final class LogHistoryManager
+{
+    private final Path masterLogFile;
+    private final long maxTotalSize;
+
+    @GuardedBy("this")
+    private long totalSize;
+    @GuardedBy("this")
+    private final PriorityQueue<LogFile> files = new PriorityQueue<>();
+
+    public LogHistoryManager(Path masterLogFile, DataSize maxTotalSize)
+    {
+        requireNonNull(masterLogFile, "masterLogFile is null");
+        requireNonNull(maxTotalSize, "maxTotalSize is null");
+
+        this.masterLogFile = masterLogFile;
+        this.maxTotalSize = maxTotalSize.toBytes();
+
+        // list existing logs
+        try {
+            Files.list(masterLogFile.getParent())
+                    .map(this::createLogFile)
+                    .filter(Optional::isPresent)
+                    .map(Optional::get)
+                    .forEach(files::add);
+            totalSize = files.stream()
+                    .mapToLong(LogFile::getSize)
+                    .sum();
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException("Unable to list existing log files", e);
+        }
+        pruneLogFilesIfNecessary(0);
+    }
+
+    public synchronized long getTotalSize()
+    {
+        return totalSize;
+    }
+
+    public synchronized Set<LogFileName> getFiles()
+    {
+        return files.stream()
+                .map(LogFile::getLogFileName)
+                .collect(toImmutableSet());
+    }
+
+    public synchronized void pruneLogFilesIfNecessary(long otherDataSize)
+    {
+        while (totalSize + otherDataSize > maxTotalSize) {
+            LogFile logFile = files.poll();
+            if (logFile == null) {
+                break;
+            }
+            totalSize -= logFile.getSize();
+            try {
+                Files.deleteIfExists(logFile.getPath());
```

Comment:
Add a comment why
- you allow file not to exist (eg file could be deleted externally and we should not fail on this)
- you subtract from `totalSize` even if file did not exist (eg files deleted externally still count towards size quota, until we decide to remove them)